src/libgap-api.c

```diff
@@ -42,46 +68,135 @@ Obj GAP_ValueGlobalVariable(const char * name)
     }
 }
 
-//
-// Evaluate a string of GAP commands
-//
-// To see an example of how to use this function
-// see tst/testlibgap/basic.c
-//
-Obj GAP_EvalString(const char * cmd)
+
+////
+//// arithmetic
+////
+
+int GAP_EQ(Obj a, Obj b)
 {
-    Obj instream;
-    Obj res;
-    Obj viewObjFunc, streamFunc;
+    return EQ(a, b);
+}
 
-    streamFunc = GAP_ValueGlobalVariable("InputTextString");
-    viewObjFunc = GAP_ValueGlobalVariable("ViewObj");
+int GAP_LT(Obj a, Obj b)
+{
+    return LT(a, b);
+}
 
-    instream = DoOperation1Args(streamFunc, MakeString(cmd));
-    res = READ_ALL_COMMANDS(instream, False, True, viewObjFunc);
-    return res;
+int GAP_IN(Obj a, Obj b)
+{
+    return IN(a, b);
+}
+
+Obj GAP_SUM(Obj a, Obj b)
+{
+    return SUM(a, b);
+}
+
+Obj GAP_DIFF(Obj a, Obj b)
+{
+    return DIFF(a, b);
+}
+
+Obj GAP_PROD(Obj a, Obj b)
+{
+    return PROD(a, b);
+}
+
+Obj GAP_QUO(Obj a, Obj b)
+{
+    return QUO(a, b);
+}
+
+Obj GAP_LQUO(Obj a, Obj b)
+{
+    return LQUO(a, b);
+}
+
+Obj GAP_POW(Obj a, Obj b)
+{
+    return POW(a, b);
+}
+
+Obj GAP_COMM(Obj a, Obj b)
+{
+    return COMM(a, b);
+}
+
+Obj GAP_MOD(Obj a, Obj b)
+{
+    return MOD(a, b);
+}
+
+
+////
+//// booleans
+////
+
+Obj GAP_True;
+Obj GAP_False;
+Obj GAP_Fail;
+
+
+////
+//// lists
+////
+
+int GAP_IsList(Obj obj)
+{
+    return obj && IS_LIST(obj);
+}
+
+Int GAP_LenList(Obj obj)
+{
+    return LEN_LIST(obj);
+}
+
+void GAP_AssList(Obj list, UInt pos, Obj val)
+{
+    if (val)
+        ASS_LIST(list, pos, val);
+    else
+        UNB_LIST(list, pos);
+}
+
+Obj GAP_ElmList(Obj list, UInt pos)
+{
+    return ELM0_LIST(list, pos);
+}
+
+Obj GAP_NewPlist(Int capacity)
+{
+    return NEW_PLIST(T_PLIST_EMPTY, capacity);
+}
+
+
+////
+//// strings
+////
+
+// Returns 1 if <obj> is a GAP string, 0 if not.
+int GAP_IsString(Obj obj)
+{
+    return obj && IS_STRING_REP(obj);
 }
 
-//
-// Returns the GAP object containing
-// <string>
 Obj GAP_MakeString(const char * string)
 {
     return MakeString(string);
 }
 
-//
-// Returns a pointer to the
-// contents of the GAP string object <string>
-char * GAP_CSTR_STRING(Obj string)
+Obj GAP_MakeImmString(const char * string)
 {
-    return CSTR_STRING(string);
+    return MakeImmString(string);
 }
 
 //
-// Returns a new empty plain list
-// with capacity <capacity>
-Obj GAP_NewPlist(Int capacity)
+// Returns a pointer to the contents of the GAP string object <string>,
+// or returns 0 if <string> is not a GAP string object.
+char * GAP_CSTR_STRING(Obj string)
```

Comment:
Isn't this very dangerous? If anything causes a garbage collection after one retrieved this pointer, it will just point *somewhere*. Also the memory area pointed to can be modified in whichever way the user wants.

Maybe sometimes this might be exactly what one wants (when?), but one might just copy the string out and not have an API function haphazardly grant access to GAP objects; 

I had a bit of a fiddle with code that one passes a buffer to (for which one might need to determine the size beforehand using a `GAP_Length` function). It feels awkward as well, the question here really is which usecases one has.

Reply:
It is as dangerous as having `Obj *bla` in the code calling libgap, but this kind of use is to be combined with a GC callback, which would go over such pointers in the code and mark bags they point to. (This is what Sagemath interface does: it has a ref-counted store of "handles" of GAP objects, and a GASMAN callback set to mark bags of stored objects).

So I'm OK with it. 

For one thing, such a function to make bags should be in libgap API.

Reply:
First off, `GAP_CSTR_STRING` is not something that *I* added, I only moved the comment to `.h` file, made it not crash on invalid argument, and documented the error behavior.

I am actually not that happy with `GAP_CSTR_STRING` myself, e.g. I don't like the name; but since it is now in a released version of GAP, I did not want to change it, which would be a breaking change after all. This is precisely the risk I warned about when I said that I'd rather be slow in adding APIs, and only want to do so after being sure somebody would, or at least might reasonably, use them; this is the case for all the APIs I added here, I created them based on inspecting the GAPJulia and SageMath code. Note that the 4.11 development cycle has only begun now, so we can still remove any and all of these new APIs before the final release, so I am not quite as concerned as with the stuff in 4.10, or possibly being backported there.

Secondly, I don't think it is any more dangerous than almost any other code directly accessing the body of an allocation in a C code base using a GC. E.g. in Julia, the C kernel also has to be aware of it. So, I think this is fine, though of course the limitations need to be documented, I'll do that ASAP.
That said, one could also implement a copying API, but that only makes sense if we also export a function to query the length of a string (which we must anyway -- actually, I thought I added one, but it seems I forgot). One thing this can't do effectively, though, is to e.g. quickly scan the content of a string for something -- though indeed it is not that clear to me whether this is all that useful in practice, but at the same time, I don't really see a practical problem with allowing it?

@dimpase That GC marking does not help with (and in fact, is irrelevant for) the issue @markuspf is thinking about (I believe): namely, that the GAP GC is a *moving* GC. So if a GC happens, the pointer returned by `GAP_CSTR_STRING` may no longer be valid.

Reply:
One thing I will say, is that holding char* pointers to GAP strings has been a source of many bugs in the GAP Kernel (i went through and fixed many such bugs some time ago) -- this is one of the few commonly used GAP functions which returns a pointer which must not be held over a GC.

Reply:
oops, thanks for reminding about GC moving... 😰 
It might partially explain trouble I am having with Sage’s interface to libgap.

Reply:
Sorry, I didn't mean to push it, I meant to just leave a suggestion (which wasn't a good suggestion, as it didn't compile of course..)

I would prefer const char*, because in my experience of GAP internals, over the years we have been very bad at keeping these char*s too long. If they are const at least the worst is a corrupted reading of a string, instead of randomly corrupting a GAP object which the pointer points to.

Reply:
On a more minor note, I wonder if it would be better to put no docs in the .c file, rather than partial docs when we probably want people to go read the full doc in the .h file.

Reply:
@dimpase FWIW there is a bug in your current code related to that, or possibly in my updates to your code.  In the error handler I made sure when using `CSTR_STRING` to immediately copy the referenced C string (our `char_to_str` function does this implicitly).  So I call `char_to_str` immediately, while the GAP string still has a reference to it.  Afterwards it may go away.

Reply:
I am perfectly happy to change the name btw, I don't think we have committed to a `libgap` API at all yet.

Reply:
@ChrisJefferson re comment, I agree, In fact I actually already removed all non-grouping comments in the .c file (well, except for one I overlooked; now also removed)